### PR TITLE
Add $unset and $geoNear stages

### PR DESCRIPTION
**Pull request by katcharov**
Also use Java tests instead of groovy.

JAVA-4594

### CODE REVIEW

File: driver-scala/src/main/scala/org/mongodb/scala/model/Aggregates.scala [hunk 1]
```diff
@@ -17,12 +17,14 @@
 package org.mongodb.scala.model
 
 import com.mongodb.client.model.fill.FillOutputField
+
 import scala.collection.JavaConverters._
-import com.mongodb.client.model.{ Aggregates => JAggregates }
+import com.mongodb.client.model.{ Aggregates => JAggregates, GeoNearOption }
```
Comment (rozza) [inline driver-scala/src/main/scala/org/mongodb/scala/model/Aggregates.scala:22]: No need to import `GeoNearOption` as you have the type alias defined in `org.mongodb.scala.model.package.scala`

**Review by rozza — CHANGES_REQUESTED** [review]
Looks good - found a few minor nits

File: driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    public Document toDocument(final BsonDocument bsonDocument) {
+        return getDefaultCodecRegistry().get(Document.class).decode(bsonDocument.asBsonReader(), DecoderContext.builder().build());
+    }
+
+    @Test
+    public void testUnset() {
+        assumeTrue(serverVersionAtLeast(4, 2));
+        insertAll("[\n"
```
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/test/functional/com/mongodb/client/model/AggregatesTest.java:92]: Keeping the `\n` allows opt-enter > copy concat to clipboard (or edit in separate window) while preserving linebreaks. Otherwise everything ends up on one line if someone needs to edit or test something.

File: driver-core/src/main/com/mongodb/client/model/GeoNearOption.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+package com.mongodb.client.model;
+
+import org.bson.Document;
+
+/**
+ * Optional fields for the {@link Aggregates#geoNear} pipeline stage.
+ *
+ * @since 4.8
+ */
+public final class GeoNearOption {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/GeoNearOption.java:26]: This seems like a new pattern for Aggregates builder methods.  Looking at what we have currently, there is BucketOptions, BucketAutoOptions, GraphLookupOptions, UnwindOptions, MergeOptions, etc.  all of which have a different design.

What's the rationale for diverging here?

**Review by rozza — CHANGES_REQUESTED** [review]

File: driver-core/src/main/com/mongodb/client/model/GeoNearOption.java (hunk 3 above)
Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/GeoNearOption.java]: Updated

File: driver-core/src/main/com/mongodb/client/model/Aggregates.java [hunk 4]
```diff
@@ -920,6 +924,48 @@ public static Bson searchMeta(final SearchCollector collector, final SearchOptio
...
+    /**
+     * Creates an $unset pipeline stage that removes/excludes fields from documents
+     *
+     * @param fields the fields to exclude. May use dot notation.
+     * @return the $unset pipeline stage
+     * @mongodb.driver.manual reference/operator/aggregation/project/ $unset
+     * @mongodb.server.release 4.2
+     * @since 4.8
+     */
+    public static Bson unset(final String... fields) {
```
[the remark this answers is not in the document]
Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/Aggregates.java:937]: Updated

File: driver-core/src/main/com/mongodb/client/model/SimplePipelineStage.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model;
+
+import org.bson.BsonDocument;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.conversions.Bson;
+
+class SimplePipelineStage implements Bson {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/SimplePipelineStage.java:23]: I'd be more inclined to revert this change and just add

>     <suppress checks="FileLength" files="Aggregates"/>

to `config/checkstyle/suppressions.xml`.  Just because it's hard to justify why this one should move but not all the others.

Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/SimplePipelineStage.java]: Updated

File: driver-core/src/main/com/mongodb/client/model/Aggregates.java [hunk 6]
```diff
@@ -920,6 +925,91 @@ public static Bson searchMeta(final SearchCollector collector, final SearchOptio
...
+            final Point near,
+            final String distanceField,
+            final GeoNearOptions options) {
+        notNull("near", near);
+        notNull("distanceField", distanceField);
+        notNull("options", options);
+        return new Bson() {
+            @Override
+            public <TDocument> BsonDocument toBsonDocument(final Class<TDocument> documentClass, final CodecRegistry codecRegistry) {
+                Document d = new Document();
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/Aggregates.java:980]: Any reason not to follow the pattern in BucketStage, where it uses BsonDocumentWriter directly?

Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/Aggregates.java]: Updated

**Review by jyemin — APPROVED** [review]
LGTM

File: driver-core/src/main/com/mongodb/client/model/Aggregates.java [hunk 7]
```diff
@@ -920,6 +924,104 @@ public static Bson searchMeta(final SearchCollector collector, final SearchOptio
...
+                });
+
+                writer.writeEndDocument();
+                writer.writeEndDocument();
+                return writer.getDocument();
+            }
+
+            @Override
+            public String toString() {
+                return "Stage{name='$densify'"
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/client/model/Aggregates.java:1000]: Nit: $geoNear

File: driver-core/src/test/functional/com/mongodb/OperationFunctionalSpecification.groovy [hunk 8]
```diff
@@ -75,12 +75,20 @@ import static com.mongodb.internal.operation.OperationUnitSpecification.getMaxWi
 
 class OperationFunctionalSpecification extends Specification {
 
-    def setup() {
+    void setup() {
```
Comment (rozza) [inline driver-core/src/test/functional/com/mongodb/OperationFunctionalSpecification.groovy:78]: Looks like this change can be reverted.

**Review by rozza — CHANGES_REQUESTED** [review]
Two minor nits - then LGTM

File: driver-core/src/main/com/mongodb/client/model/Aggregates.java (hunk 7 above)
Comment (katcharov) [reply driver-core/src/main/com/mongodb/client/model/Aggregates.java]: Good eye
